telemetry/telemetry_linux.go

```diff
@@ -0,0 +1,271 @@
+			}
+
+			memFree = memFree / KB
+			memArray = append(memArray, memFree)
+		}
+	}
+
+	memInfo := &MemInfo{MemTotal: memArray[0], MemFree: memArray[1]}
+
+	return memInfo, nil
+}
+
+// This function retrieves VMs disk usage.
+func getDiskInfo(path string) (*DiskInfo, error) {
+	fs := syscall.Statfs_t{}
+	err := syscall.Statfs(path, &fs)
+	if err != nil {
+		return nil, fmt.Errorf("Statfs call failed with error %v", err)
+	}
+
+	total := fs.Blocks * uint64(fs.Bsize) / MB
+	free := fs.Bfree * uint64(fs.Bsize) / MB
+
+	diskInfo := &DiskInfo{DiskTotal: total, DiskFree: free}
+	return diskInfo, nil
+}
+
+// This function  creates a report with system details(memory, disk, cpu).
+func GetSystemDetails() (*SystemInfo, error) {
+	var cpuCount int = 0
+
+	linesArr, err := ReadFileByLines("/proc/cpuinfo")
+	if err != nil {
+		return nil, err
+	}
+
+	for _, line := range linesArr {
+		fieldArray := strings.Split(line, ":")
+		fieldArray[0] = strings.TrimSpace(fieldArray[0])
+		if strings.Compare(fieldArray[0], "processor") == 0 {
+			cpuCount += 1
+		}
+	}
+
+	memInfo, err := getMemInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	diskInfo, err := getDiskInfo("/")
+	if err != nil {
+		return nil, err
+	}
+
+	sysInfo := &SystemInfo{
+		MemVMTotal:  memInfo.MemTotal,
+		MemVMFree:   memInfo.MemFree,
+		DiskVMTotal: diskInfo.DiskTotal,
+		DiskVMFree:  diskInfo.DiskFree,
+		CPUCount:    cpuCount,
+	}
+
+	return sysInfo, nil
+}
+
+// This function  creates a report with os details(ostype, version).
+func GetOSDetails() (*OSInfo, error) {
+	osType := "Linux"
+
+	linesArr, err := ReadFileByLines("/etc/issue")
+	if err != nil {
+		return nil, err
+	}
+
+	osInfoArr := strings.Split(linesArr[0], " ")
+
+	out, err := exec.Command("uname", "-r").Output()
+	kernelVersion := string(out)
+
+	osDetails := &OSInfo{OSType: osType, OSVersion: osInfoArr[1], KernelVersion: kernelVersion, OSDistribution: osInfoArr[0]}
+	return osDetails, nil
+}
+
+// This function  creates a report with interface details(ip, mac, name, secondaryca count).
+func GetInterfaceDetails(queryUrl string) (*InterfaceInfo, error) {
+
+	var (
+		macAddress       string
+		secondaryCACount int
+		primaryCA        string
+		subnet           string
+		ifName           string
+	)
+
+	if queryUrl == "" {
+		return nil, fmt.Errorf("IpamQueryUrl is null")
+	}
+
+	interfaces, err := net.Interfaces()
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := http.Get(queryUrl)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	// Decode XML document.
+	var doc xmlDocument
+	decoder := xml.NewDecoder(resp.Body)
+	err = decoder.Decode(&doc)
+	if err != nil {
+		return nil, err
+	}
+
+	// For each interface...
+	for _, i := range doc.Interface {
+		i.MacAddress = strings.ToLower(i.MacAddress)
+
+		if i.IsPrimary {
+			// Find the interface with the matching MacAddress.
+			for _, iface := range interfaces {
+				macAddr := strings.Replace(iface.HardwareAddr.String(), ":", "", -1)
```

Comment:
Should we keep the colons for readability?

Reply:
The wireserver returns MAC without colon. So we need to remove colon for that. Anyway in report the mac will be sent in colon format.

Reply:
nice.